command/oauth/cmd.go

```diff
@@ -520,29 +559,53 @@ func newOauth(provider, clientID, clientSecret, authzEp, tokenEp, scope, prompt
 		}, nil
 	default:
 		userinfoEp := ""
-		if authzEp == "" && tokenEp == "" {
-			d, err := disco(provider)
-			if err != nil {
-				return nil, err
-			}
-
-			if _, ok := d["authorization_endpoint"]; !ok {
-				return nil, errors.New("missing 'authorization_endpoint' in provider metadata")
+		if opts.Device {
+			if deviceAuthzEp == "" && tokenEp == "" {
+				d, err := disco(provider)
+				if err != nil {
+					return nil, err
+				}
+
+				if _, ok := d["device_authorization_endpoint"]; !ok {
+					return nil, errors.New("missing 'device_authorization_endpoint' in provider metadata")
+				}
+				if _, ok := d["token_endpoint"]; !ok {
+					return nil, errors.New("missing 'token_endpoint' in provider metadata")
+				}
+				deviceAuthzEp = d["device_authorization_endpoint"].(string)
+				tokenEp = d["token_endpoint"].(string)
+				userinfoEp = d["token_endpoint"].(string)
 			}
-			if _, ok := d["token_endpoint"]; !ok {
-				return nil, errors.New("missing 'token_endpoint' in provider metadata")
+		} else {
+			if authzEp == "" && tokenEp == "" {
+				d, err := disco(provider)
+				if err != nil {
+					return nil, err
+				}
+
+				if _, ok := d["authorization_endpoint"]; !ok {
+					return nil, errors.New("missing 'authorization_endpoint' in provider metadata")
+				}
+				if _, ok := d["token_endpoint"]; !ok {
+					return nil, errors.New("missing 'token_endpoint' in provider metadata")
+				}
+				if _, ok := d["device_authorization_endpoint"]; !ok {
+					return nil, errors.New("missing 'token_endpoint' in provider metadata")
+				}
+				authzEp = d["authorization_endpoint"].(string)
+				tokenEp = d["token_endpoint"].(string)
+				userinfoEp = d["token_endpoint"].(string)
```

Comment:
We can refactor this into one if condition, failing on different cases depending on the value of ops.Device.